### PR TITLE
dtv: Remove manual memory management

### PR DESCRIPTION
**Pull request by ThomasHabets**
All removed except some tricky stuff in `dvb_ldpc_bb_impl`.

### CODE REVIEW

File: gr-dtv/lib/atsc/atsc_deinterleaver_impl.cc [hunk 1]
```diff
@@ -30,25 +30,22 @@ atsc_deinterleaver_impl::atsc_deinterleaver_impl()
                      io_signature::make(1, 1, sizeof(atsc_mpeg_packet_rs_encoded))),
       alignment_fifo(156)
 {
-    m_fifo.resize(52);
+    m_fifo.reserve(s_interleavers);
 
-    for (int i = 0; i < 52; i++)
-        m_fifo[52 - 1 - i] = new interleaver_fifo<unsigned char>(i * 4);
+    for (int i = 0; i < s_interleavers; i++)
+        m_fifo.emplace_back((52 - 1 - i) * 4);
```
Comment (jdemel) [inline gr-dtv/lib/atsc/atsc_deinterleaver_impl.cc:36]: Is `s_interleaver` a constant? aka `s_interleaver = 52;`? In that case, I'd suggest to use the constant consistently, I assume it would be correct to replace `52` with `s_interleaver` as well. Also, the other literals might be good candidates for replacement too. Well, only `4`.

Found it: `static constexpr int s_interleavers = 52;`. Since this is a constant, would it be better to spell it `S_INTERLEAVERS`?

File: gr-dtv/lib/atsc/interleaver_fifo.h [hunk 2]
```diff
@@ -22,52 +23,40 @@ class interleaver_fifo
 {
 public:
     interleaver_fifo(unsigned int size);
```
Comment (jdemel) [inline gr-dtv/lib/atsc/interleaver_fifo.h:25]: What is the difference between this FIFO and the one in `atsc_interleaver_fifo.h`? Can they be merged into one?

File: gr-dtv/lib/dvbt/dvbt_map_impl.h [hunk 3]
```diff
@@ -20,20 +20,20 @@ class dvbt_map_impl : public dvbt_map
...
-    unsigned char d_alpha;
+    const unsigned char d_alpha;
     // Gain for the complex values
-    float d_gain;
+    const float d_gain;
 
-    gr_complex* d_constellation_points;
+    std::vector<gr_complex> d_constellation_points;
 
     void make_constellation_points(int size, int step, int alpha);
```
Comment (jdemel) [inline gr-dtv/lib/dvbt/dvbt_map_impl.h:38]: What's the difference betwenn this function and the one in `dvbt_demap`? Would it be possible to have one constellation object for both?

File: gr-dtv/lib/dvbt/dvbt_viterbi_decoder_impl.h [hunk 4]
```diff
@@ -101,13 +103,13 @@ class dvbt_viterbi_decoder_impl : public dvbt_viterbi_decoder
     int mettab[2][256];
```
Comment (jdemel) [inline gr-dtv/lib/dvbt/dvbt_viterbi_decoder_impl.h:103]: I know this has fixed size, but  for consistency, wouldn't it be better to use a `std::vector` here as well?

File: gr-dtv/lib/dvbt2/dvbt2_paprtr_cc_impl.cc [hunk 5]
```diff
@@ -808,22 +736,26 @@ int dvbt2_paprtr_cc_impl::work(int noutput_items,
...
-                            ctemp, ones_freq, result, papr_fft_size);
-                        volk_32f_x2_subtract_32f(
-                            (float*)c, (float*)c, (float*)ctemp, papr_fft_size * 2);
-                        memcpy(r, rNew, sizeof(gr_complex) * N_TR);
+                            ctemp.data(), ones_freq.data(), result, papr_fft_size);
+                        volk_32f_x2_subtract_32f((float*)c.data(),
+                                                 (float*)c.data(),
+                                                 (float*)ctemp.data(),
+                                                 papr_fft_size * 2);
+                        std::copy(std::begin(rNew), std::end(rNew), std::begin(r));
```
Comment (jdemel) [inline gr-dtv/lib/dvbt2/dvbt2_paprtr_cc_impl.cc:755]: So `std::copy` translates to a `memcpy`? Or does it incur more overhead? I've seen some `memcpy` in the other files, that might want to be replaced as well.

File: gr-dtv/lib/dvbt2/dvbt2_paprtr_cc_impl.h [hunk 6]
```diff
@@ -94,6 +94,11 @@ class dvbt2_paprtr_cc_impl : public dvbt2_paprtr_cc
                          float vclip,
                          int iterations,
                          unsigned int vlength);
+
+    // Disable copy because raw pointers.
+    dvbt2_paprtr_cc_impl(const dvbt2_paprtr_cc_impl&) = delete;
```
Comment (jdemel) [inline gr-dtv/lib/dvbt2/dvbt2_paprtr_cc_impl.h:99]: Which members are raw pointers? Do you refer to `p2_papr_map_*k`?

**Review by jdemel — COMMENTED** [review]
I have a few comments. Besides this looks good. It's amazing how modern c++ enables one to do more with fewer lines of code.

**Review by willcode — APPROVED** [review]
Great improvement. If it works and performance is similar or better, we could come back and make more tweaks later.

File: gr-dtv/lib/atsc/atsc_deinterleaver_impl.cc (hunk 1 above)
Comment (ThomasHabets) [reply gr-dtv/lib/atsc/atsc_deinterleaver_impl.cc:36]: Static member variables [should be prefixed with `s_`](https://github.com/gnuradio/greps/blob/master/grep-0001-coding-guidelines.md#code-formatting-guidelines).

I could make it a global variable in `gr::dtv` and capitalize it, but there's no existing namespace outside the class for just atsc.

File: gr-dtv/lib/atsc/interleaver_fifo.h (hunk 2 above)
Comment (ThomasHabets) [reply gr-dtv/lib/atsc/interleaver_fifo.h:25]: I don't know. I'd say that would make for a separate PR, though.

File: gr-dtv/lib/dvbt/dvbt_map_impl.h (hunk 3 above)
Comment (ThomasHabets) [reply gr-dtv/lib/dvbt/dvbt_map_impl.h:38]: Same here. Looks like there more simplifying work to do in `dtv` that isn't about removing manual memory management.

File: gr-dtv/lib/dvbt/dvbt_viterbi_decoder_impl.h (hunk 4 above)
Comment (ThomasHabets) [reply gr-dtv/lib/dvbt/dvbt_viterbi_decoder_impl.h:103]: Maybe `std::array<>` would be even better.

I'd say so, for another PR. :-)

File: gr-dtv/lib/dvbt2/dvbt2_paprtr_cc_impl.cc (hunk 5 above)
Comment (ThomasHabets) [reply gr-dtv/lib/dvbt2/dvbt2_paprtr_cc_impl.cc:755]: `std::copy` is like `memcpy` but actually knows what it's copying. `memcpy` has to be told how many bytes, which makes it easy to get something wrong, as types change and when copy-pasting.

Usually template implementations actually compile down to faster than simple functions (because the compiler has type access . Though `memcpy` is special in most compilers.

tl;dr: they'll be just as fast, and with today's compilers may even compile down to the exact same instructions. That's my default unless benchmarks prove otherwise.

File: gr-dtv/lib/dvbt2/dvbt2_paprtr_cc_impl.h (hunk 6 above)
Comment (ThomasHabets) [reply gr-dtv/lib/dvbt2/dvbt2_paprtr_cc_impl.h:99]: `p2_papr_map` and `tr_papr_map`.

But hmm… actually those point to const static lifetime objects, so are safe to copy. Removing these lines.

**Comment by ThomasHabets** [comment]
Running tests before pushing revised version.

**Comment by ThomasHabets** [comment]
Updated.

**Review by jdemel — APPROVED** [review]
LGTM.
